### PR TITLE
Skipchain unit tests

### PR DESCRIPTION
**Pull request by Gilthoniel**
This aims to add unit test to the skipchain module. This is a first wave before extracting the consensus module logic. After that the final coverage will be done.

### CODE REVIEW

File: crypto/bls/mod.go [hunk 1]
```diff
@@ -59,18 +59,20 @@ func (f publicKeyFactory) FromAny(src *any.Any) (crypto.PublicKey, error) {
...
+		err := point.UnmarshalBinary(msg.GetData())
+		if err != nil {
+			return nil, err
+		}
+
+		return publicKey{point: point}, nil
+	case *any.Any:
+		return f.fromAny(msg)
+	default:
 		return nil, errors.New("invalid public key type")
```
Comment (nkcr) [inline crypto/bls/mod.go]:
```suggestion
		return nil, xerrors.Errorf("invalid public key type '%v'", msg)
```

File: crypto/bls/mod.go [hunk 2]
```diff
@@ -86,12 +88,14 @@ func (f signatureFactory) FromAny(src *any.Any) (crypto.Signature, error) {
...
 func (f signatureFactory) FromProto(src proto.Message) (crypto.Signature, error) {
-	sigproto, ok := src.(*SignatureProto)
-	if !ok {
+	switch msg := src.(type) {
+	case *SignatureProto:
+		return signature{data: msg.GetData()}, nil
+	case *any.Any:
+		return f.FromAny(msg)
+	default:
 		return nil, errors.New("invalid signature type")
```
Comment (nkcr) [inline crypto/bls/mod.go]:
```suggestion
		return nil, xerrors.Errorf("invalid signature type '%v'", msg)
```

File: crypto/bls/mod.go [hunk 3]
```diff
@@ -59,18 +59,20 @@ func (f publicKeyFactory) FromAny(src *any.Any) (crypto.PublicKey, error) {
...
 
 func (f publicKeyFactory) FromProto(src proto.Message) (crypto.PublicKey, error) {
-	pkp, ok := src.(*PublicKeyProto)
-	if !ok {
+	switch msg := src.(type) {
+	case *PublicKeyProto:
+		point := suite.Point()
+		err := point.UnmarshalBinary(msg.GetData())
+		if err != nil {
+			return nil, err
```
Comment (nkcr) [inline crypto/bls/mod.go]:
```suggestion
			return nil, xerrors.Errorf("failed to unmarshal msg to point: %v", err)
```

File: encoding/errors.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	expectedType := reflect.TypeOf(expected)
+
+	return TypeError{
+		curr:     currType.String(),
+		expected: expectedType.String(),
+	}
+}
+
+func (e TypeError) Error() string {
+	return fmt.Sprintf("got message type %#v but expect %#v", e.curr, e.expected)
```
Comment (nkcr) [inline encoding/errors.go]:
```suggestion
	return fmt.Sprintf("got message type '%#v' but expected '%#v'", e.curr, e.expected)
```

File: blockchain/skipchain/block.go [hunk 5]
```diff
@@ -217,13 +244,17 @@ type Chain []SkipBlock
...
+	}
 
 	proof := Proof{
-		ForwardLinks: make([]ForwardLink, len(chain)),
+		GenesisBlock: c[0],
+		ForwardLinks: make([]ForwardLink, numForwardLink),
 	}
 
-	for i, block := range chain {
+	for i, block := range c[:numForwardLink] {
```
Comment (nkcr) [inline blockchain/skipchain/block.go]:
```suggestion
	for i, block := range c {
```

File: blockchain/skipchain/block.go [hunk 6]
```diff
@@ -217,13 +244,17 @@ type Chain []SkipBlock
...
-	chain := c[:len(c)-1]
+	numForwardLink := int(math.Max(0, float64(len(c)-1)))
+	if len(c) == 0 {
+		return Proof{}
+	}
 
 	proof := Proof{
-		ForwardLinks: make([]ForwardLink, len(chain)),
+		GenesisBlock: c[0],
+		ForwardLinks: make([]ForwardLink, numForwardLink),
```
Comment (nkcr) [inline blockchain/skipchain/block.go]:
```suggestion
		ForwardLinks: make([]ForwardLink, len(c)),
```

File: blockchain/skipchain/block.go [hunk 7]
```diff
@@ -262,26 +293,54 @@ func (p Proof) Pack() (proto.Message, error) {
...
-			if !bytes.Equal(prev, link.From) {
-				return xerrors.Errorf("got previous block %x but expect %x in forward link",
-					link.From, prev)
-			}
-		}
+		prev = link.GetTo()
+	}
+
+	if !bytes.Equal(prev, block.GetID()) {
+		return xerrors.Errorf("got forward link to %v but expect %v", prev, block.GetID())
```
Comment (nkcr) [inline blockchain/skipchain/block.go]:
```suggestion
		return xerrors.Errorf("got forward link to '%v' but expected '%v'", prev, block.GetID())
```

File: blockchain/skipchain/block.go [hunk 8]
```diff
@@ -262,26 +293,54 @@ func (p Proof) Pack() (proto.Message, error) {
...
+}
+
+func (p Proof) verifyLink(v crypto.Verifier, pubkeys []crypto.PublicKey, link ForwardLink, prev BlockID) error {
+	err := link.Verify(v, pubkeys)
+	if err != nil {
+		return xerrors.Errorf("couldn't verify the signatures: %w", err)
+	}
+
+	if !bytes.Equal(prev, link.GetFrom()) {
+		return xerrors.Errorf("got previous block %v but expect %v in forward link",
```
Comment (nkcr) [inline blockchain/skipchain/block.go]:
```suggestion
		return xerrors.Errorf("got previous block '%v' but expected '%v' in forward link",
```

File: blockchain/skipchain/block.go (hunk 5 above)
Comment (Gilthoniel) [reply blockchain/skipchain/block.go]: there is one forward link less than the number of blocks

**Pull request merged**